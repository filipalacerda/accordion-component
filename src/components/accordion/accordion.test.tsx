import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import Accordion from ".";
import userEvent from "@testing-library/user-event";

describe("Accordion", () => {
  const title = "Title 1";
  const content = "This is the content";

  describe("when component renders", () => {
    let container: HTMLElement;

    beforeEach(() => {
      container = render(
        <Accordion title={title} content={content} />
      ).container;
    });

    it("should render the component", () => {
      expect(container).toMatchSnapshot();
    });

    it("should render the provided title", () => {
      expect(screen.getByText(title)).toBeInTheDocument();
    });

    it("should render the open accordion icon", () => {
      expect(screen.getByTestId("arrow-down")).toBeInTheDocument();
    });

    it("should render a closed accordion", () => {
      expect(screen.queryByTestId("accordion-content")).not.toBeInTheDocument();
    });

    it("should not render the close accordion icon", () => {
      expect(screen.queryByTestId("arrow-up")).not.toBeInTheDocument();
    });
  });

  describe("with disabled property", () => {
    it("should not open the accordion when user clicks", async () => {
      render(<Accordion title={title} content={content} isDisabled />);

      expect(screen.queryByTestId("arrow-up")).not.toBeInTheDocument();
      await userEvent.click(screen.getByTestId("accordion-button"));

      expect(screen.queryByTestId("arrow-up")).not.toBeInTheDocument();
      expect(screen.queryByTestId("accordion-content")).not.toBeInTheDocument();
      expect(screen.getByTestId("arrow-down")).toBeInTheDocument();
    });
  });

  describe("with defaultExpanded", () => {
    it("should render the accordion opened", () => {
      render(<Accordion title={title} content={content} defaultExpanded />);

      // Open accordion icon
      expect(screen.queryByTestId("arrow-down")).not.toBeInTheDocument();
      // Close accordion icon
      expect(screen.getByTestId("arrow-up")).toBeInTheDocument();

      expect(screen.getByTestId("accordion-content")).toBeInTheDocument();
    });
  });

  describe("keydown events", () => {
    let container: HTMLElement;
    let button: HTMLElement;

    beforeEach(() => {
      container = render(
        <Accordion title={title} content={content} />
      ).container;

      button = screen.getByTestId("accordion-button");
    });

    describe("when the user presses arrowDown key", () => {
      it("should open the accordion", async () => {
        await fireEvent.keyDown(button, {
          key: "ArrowDown",
          code: "ArrowDown",
        });

        expect(screen.getByTestId("accordion-content")).toBeInTheDocument();
      });
    });

    describe("when the user presses ArrowDown key", () => {
      it("should close the accordion", async () => {
        // Open the accordion
        await fireEvent.keyDown(button, {
          key: "ArrowDown",
          code: "ArrowDown",
        });

        expect(screen.getByTestId("accordion-content")).toBeInTheDocument();

        // close the accordion

        await fireEvent.keyDown(button, {
          key: "ArrowUp",
        });
        expect(
          screen.queryByTestId("accordion-content")
        ).not.toBeInTheDocument();
      });
    });
  });
});
