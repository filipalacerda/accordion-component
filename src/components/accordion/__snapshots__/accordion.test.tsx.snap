// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion when component renders should render the component 1`] = `
<div>
  <section
    class="css-nvtxod"
    data-testid="accordion"
  >
    <button
      aria-controls="accordion-content"
      aria-expanded="false"
      class="css-1054hxc"
      data-testid="accordion-button"
    >
      Title 1
      <svg
        class="css-me3p27"
        data-testid="arrow-down"
        fill="currentColor"
        height="24"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          Open accordion icon
        </title>
        <path
          d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
        />
      </svg>
    </button>
  </section>
</div>
`;
